File: src/morphic/converter/scanner.py
"""
Folder scanner – walks directories and inventories image/video files
for the format converter.
"""

from __future__ import annotations

import os
from pathlib import Path

from morphic.converter.constants import IMAGE_CONVERSIONS, VIDEO_CONVERSIONS
from morphic.shared.constants import IMAGE_EXTENSIONS, VIDEO_EXTENSIONS
from morphic.shared.utils import normalise_ext


def get_compatible_targets(source_path: str) -> list[str]:
    """Return list of extensions the source can be converted to."""
    ext = normalise_ext(Path(source_path).suffix)
    if ext in IMAGE_CONVERSIONS:
        return IMAGE_CONVERSIONS[ext]
    if ext in VIDEO_CONVERSIONS:
        return VIDEO_CONVERSIONS[ext]
    return []


def scan_folder(
    folder: str,
    include_subfolders: bool = True,
    filter_type: str = "both",
) -> dict:
    """
    Walk *folder* and return a summary + full file list.

    Parameters
    ----------
    folder : str
        Absolute path to the directory to scan.
    include_subfolders : bool
        Whether to recurse into subdirectories.
    filter_type : str
        One of ``"images"``, ``"videos"``, ``"both"``.

    Returns
    -------
    dict
        ``{"folder", "summary", "files", "total"}``
    """
    allowed: set[str] = set()
    if filter_type in ("images", "both"):
        allowed |= IMAGE_EXTENSIONS
    if filter_type in ("videos", "both"):
        allowed |= VIDEO_EXTENSIONS

    summary: dict[str, int] = {}
    files: list[dict] = []

    if include_subfolders:
        walker = os.walk(folder)
    else:
        try:
            entries = os.listdir(folder)
        except PermissionError:
            entries = []
        walker = [(folder, [], entries)]  # type: ignore[assignment]

    for dirpath, _dirs, filenames in walker:
        for fname in filenames:
            full = os.path.join(dirpath, fname)
            ext = normalise_ext(Path(fname).suffix)
            if ext not in allowed:
                continue
            try:
                size = os.path.getsize(full)
            except OSError:
                size = 0

            ftype = "image" if ext in IMAGE_EXTENSIONS else "video"
            targets = get_compatible_targets(full)

            summary[ext] = summary.get(ext, 0) + 1
            files.append({
                "path": full,
                "name": fname,
                "ext": ext,
                "size": size,
                "type": ftype,
                "targets": targets,
            })

    sorted_summary = dict(sorted(summary.items(), key=lambda x: -x[1]))

    return {
        "folder": folder,
        "summary": sorted_summary,
        "files": sorted(files, key=lambda f: f["name"].lower()),
        "total": len(files),
    }

File: internal/converter/scanner.go
package converter

import (
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/exterex/morphic/internal/shared"
)

// ScanResult holds the output of a folder scan.
type ScanResult struct {
	Folder  string            `json:"folder"`
	Summary map[string]int    `json:"summary"`
	Files   []FileEntry       `json:"files"`
	Total   int               `json:"total"`
}

// FileEntry describes one file found during scanning.
type FileEntry struct {
	Path    string   `json:"path"`
	Name    string   `json:"name"`
	Ext     string   `json:"ext"`
	Size    int64    `json:"size"`
	Type    string   `json:"type"`
	Targets []string `json:"targets"`
}

// ScanFolder walks folder and returns an inventory of convertible media files.
func ScanFolder(folder string, includeSubfolders bool, filterType string) (*ScanResult, error) {
	allowed := make(map[string]struct{})
	if filterType == "images" || filterType == "both" {
		for k, v := range shared.ImageExtensions {
			allowed[k] = v
		}
	}
	if filterType == "videos" || filterType == "both" {
		for k, v := range shared.VideoExtensions {
			allowed[k] = v
		}
	}

	summary := make(map[string]int)
	var files []FileEntry

	if includeSubfolders {
		filepath.WalkDir(folder, func(path string, d os.DirEntry, err error) error {
			if err != nil {
				return nil
			}
			if d.IsDir() {
				name := strings.ToLower(d.Name())
				if _, excl := shared.ExcludedFolders[name]; excl {
					return filepath.SkipDir
				}
				return nil
			}
			addFileEntry(path, d, allowed, summary, &files)
			return nil
		})
	} else {
		entries, err := os.ReadDir(folder)
		if err != nil {
			return nil, err
		}
		for _, d := range entries {
			if d.IsDir() {
				continue
			}
			path := filepath.Join(folder, d.Name())
			addFileEntry(path, d, allowed, summary, &files)
		}
	}

	// Sort summary by count descending
	type kv struct {
		K string
		V int
	}
	var sorted []kv
	for k, v := range summary {
		sorted = append(sorted, kv{k, v})
	}
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].V > sorted[j].V })
	sortedSummary := make(map[string]int, len(sorted))
	for _, s := range sorted {
		sortedSummary[s.K] = s.V
	}

	// Sort files by name
	sort.Slice(files, func(i, j int) bool {
		return strings.ToLower(files[i].Name) < strings.ToLower(files[j].Name)
	})

	return &ScanResult{
		Folder:  folder,
		Summary: sortedSummary,
		Files:   files,
		Total:   len(files),
	}, nil
}

func addFileEntry(path string, d os.DirEntry, allowed map[string]struct{}, summary map[string]int, files *[]FileEntry) {
	ext := shared.NormaliseExt(strings.ToLower(filepath.Ext(path)))
	if _, ok := allowed[ext]; !ok {
		return
	}

	var size int64
	if info, err := d.Info(); err == nil {
		size = info.Size()
	}

	ftype := "image"
	if _, ok := shared.VideoExtensions[ext]; ok {
		ftype = "video"
	}

	targets := GetCompatibleTargets(path)
	summary[ext]++
	*files = append(*files, FileEntry{
		Path:    path,
		Name:    d.Name(),
		Ext:     ext,
		Size:    size,
		Type:    ftype,
		Targets: targets,
	})
}
